File: backend/app/adapters/pricing.py
from app.schemas.order import OrderItemResponse


def resolve_price(product: dict, quantity: int, role: str, total_lots: int) -> float:
    """Resolve the unit price for one line item under the two-tier rule.

    Authoritative pricing rule (ADR-001, post reversion from three-tier):
      1. quantity < minimum_wholesale_lots        -> retail_price_per_lot
      2. total_lots >= 10 (any role, incl. admin) -> wholesale_price_per_lot
      3. otherwise                                -> retail_price_per_lot

    The role parameter is retained for backward compatibility with
    order_intake.resolve_all_items() but does not alter pricing.
    """
    retail = float(product["retail_price_per_lot"])
    wholesale = float(product["wholesale_price_per_lot"])

    if quantity < product["minimum_wholesale_lots"]:
        return retail

    if total_lots >= 10:
        return min(wholesale, retail)

    return retail


def resolve_all_items(products_map: dict[str, dict], checkout_items: list[dict], role: str, total_lots: int) -> list[OrderItemResponse]:
    resolved = []
    for item in checkout_items:
        product = products_map.get(item["product_id"])
        if product is None:
            continue
        unit_price = resolve_price(product, item["quantity"], role, total_lots)
        resolved.append(
            OrderItemResponse(
                product_id=item["product_id"],
                title=product["title"],
                quantity_ordered=item["quantity"],
                unit_price_applied=unit_price,
            )
        )
    return resolved

File: frontend/src/lib/pricing.ts
import type { Product } from "@/types"
import type { UserRole } from "@/types"

/**
 * Client-side mirror of the authoritative backend pricing rule.
 *
 * SOURCE OF TRUTH: backend/app/adapters/pricing.py — keep this file in sync.
 * Spec: docs/superpowers/specs/2026-08-03-three-tier-pricing-design.md
 *
 * Resolved per line item, in this order:
 *   1. quantity < minimum_wholesale_lots        -> retail_price_per_lot
 *   2. role === "wholesale_approved" (any size) -> approved_price_per_lot
 *   3. totalLots >= 10 (any role, incl. admin)  -> wholesale_price_per_lot
 *   4. otherwise                                -> retail_price_per_lot
 *
 * The product param accepts the full `Product` shape or a partial snapshot
 * (e.g. a stale cart item persisted in localStorage before the approved-price
 * column existed). When `approved_price_per_lot` is missing it falls back to
 * the wholesale price — matching the migration-009 backfill semantics. This
 * mirror is used for display/estimates only; the backend charge is always
 * authoritative.
 */
export function resolvePrice(
  product: Pick<
    Product,
    "retail_price_per_lot" | "wholesale_price_per_lot" | "minimum_wholesale_lots" | "approved_price_per_lot"
  >,
  quantity: number,
  role: UserRole | null,
  totalLots: number
): number {
  const retail = Number(product.retail_price_per_lot)
  const wholesale = Number(product.wholesale_price_per_lot)

  if (quantity < product.minimum_wholesale_lots) {
    return retail
  }

  if (role === "wholesale_approved") {
    const approved = product.approved_price_per_lot
    if (approved !== undefined && approved !== null) {
      const approvedFloat = Number(approved)
      // Mirrors of products_approved_price_check and
      // products_wholesale_price_check: never overcharge with bad config.
      if (approvedFloat <= wholesale) {
        return Math.min(approvedFloat, retail)
      }
    }
    return Math.min(wholesale, retail)
  }

  if (totalLots >= 10) {
    return Math.min(wholesale, retail)
  }

  return retail
}

export type PricingTier = "RETAIL" | "WHOLESALE" | "APPROVED"

/**
 * Returns the tier label that applies to a line given the same rule as
 * `resolvePrice`. Used to label per-line and per-cart tier indicators.
 */
export function resolveTier(
  product: Pick<
    Product,
    "retail_price_per_lot" | "wholesale_price_per_lot" | "minimum_wholesale_lots" | "approved_price_per_lot"
  >,
  quantity: number,
  role: UserRole | null,
  totalLots: number
): PricingTier {
  if (quantity < product.minimum_wholesale_lots) {
    return "RETAIL"
  }
  if (role === "wholesale_approved") {
    const approved = product.approved_price_per_lot
    if (approved !== undefined && approved !== null && Number(approved) <= Number(product.wholesale_price_per_lot)) {
      return "APPROVED"
    }
    return "WHOLESALE"
  }
  if (totalLots >= 10) {
    return "WHOLESALE"
  }
  return "RETAIL"
}
